src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java

```diff
@@ -61,82 +63,84 @@ public NtlmAuthenticator create() {
+                byte[] sessionkey;
+
+                byte[] userSessionKey = ntlmFunctions.hmac_md5(responseKeyNT, Arrays.copyOfRange(ntlmv2Response, 0, 16)); // first 16 bytes of ntlmv2Response is ntProofStr
+                EnumSet<NtlmNegotiateFlag> negotiateFlags = challenge.getNegotiateFlags();
+                if (negotiateFlags.contains(NtlmNegotiateFlag.NTLMSSP_NEGOTIATE_KEY_EXCH)
+                    && (negotiateFlags.contains(NtlmNegotiateFlag.NTLMSSP_NEGOTIATE_SIGN)
+                    || negotiateFlags.contains(NtlmNegotiateFlag.NTLMSSP_NEGOTIATE_SEAL)
+                    || negotiateFlags.contains(NtlmNegotiateFlag.NTLMSSP_NEGOTIATE_ALWAYS_SIGN))
+                    ) {
+                    byte[] masterKey = new byte[16];
+                    random.nextBytes(masterKey);
+                    sessionkey = ntlmFunctions.encryptRc4(userSessionKey, masterKey);
+                    session.setSigningKey(masterKey);
+                } else {
+                    sessionkey = userSessionKey;
+                    session.setSigningKey(sessionkey);
+                }
+
+                // If NTLM v2 is used, KeyExchangeKey MUST be set to the given 128-bit SessionBaseKey value.
+
+                // MIC (16 bytes) provided if in AvPairType is key MsvAvFlags with value & 0x00000002 is true
+                Object msvAvFlags = challenge.getAvPairObject(AvId.MsvAvFlags);
+                if (msvAvFlags != null && ((int) msvAvFlags & 0x00000002) > 0) {
+                    // MIC should be calculated
+                    NtlmAuthenticate resp = new NtlmAuthenticate(new byte[0], ntlmv2Response,
+                        context.getUsername(), context.getDomain(), null, sessionkey, EnumWithValue.EnumUtils.toLong(negotiateFlags),
+                        true
+                    );
+
+                    // TODO correct hash should be tested
+
+                    Buffer.PlainBuffer concatenatedBuffer = new Buffer.PlainBuffer(Endian.LE);
+                    concatenatedBuffer.putRawBytes(negTokenTarg.getResponseToken()); //negotiateMessage
+                    concatenatedBuffer.putRawBytes(challenge.getServerChallenge()); //challengeMessage
+                    resp.writeAutentificateMessage(concatenatedBuffer); //authentificateMessage
+
+                    byte[] mic = ntlmFunctions.hmac_md5(userSessionKey, concatenatedBuffer.getCompactData());
+                    resp.setMic(mic);
+                    return negTokenTarg(resp, negTokenTarg.getResponseToken());
+                } else {
+                    NtlmAuthenticate resp = new NtlmAuthenticate(new byte[0], ntlmv2Response,
+                        context.getUsername(), context.getDomain(), null, sessionkey, EnumWithValue.EnumUtils.toLong(negotiateFlags),
+                        false
+                    );
+                    return negTokenTarg(resp, negTokenTarg.getResponseToken());
+                }
             }
+        } catch (SpnegoException spne) {
```

Comment:
Hi Guys, Thanks for the hardwork. One question.
Just trying to understand, why in NTLMAuthentication, SpnegoException is thrown?